### PR TITLE
Move state summary cache to DB 

### PR DESCRIPTION
**Pull request by terencechain**
**What type of PR is this?**

> Clean up

**What does this PR do? Why is it needed?**

This PR moves state summary cache from its independent pkg for suppling `stateGen` and `db` pkgs to become part of `db` pkg. This efficiently eliminates separation of concerns, we no longer have to operate state summary cache in between `db` and `stateGen`. 

The design of state summary cache in db is quite simple, we track the cached state summary count and when it reaches a prune threshold, we dump the cached state summaries to the DB and clears up the cache.

The diff is bad due to test setups. I will narrative on the diffs on which to review.

**Which issues(s) does this PR fix?**

Part of #8098 

**Other notes for review**

Tested it run time by syncing and validating

### CODE REVIEW

File: beacon-chain/blockchain/service.go [hunk 1]
```diff
@@ -342,11 +342,6 @@ func (s *Service) Stop() error {
 		}
 	}
 
-	// Save cached state summaries to the DB before stop.
```
Comment (terencechain) [inline beacon-chain/blockchain/service.go:345]: Instead, we save cached state summaries in DB pkg's `Close()`

File: beacon-chain/db/kv/kv.go [hunk 2]
```diff
@@ -147,6 +147,13 @@ func (s *Store) ClearDB() error {
 // Close closes the underlying BoltDB database.
 func (s *Store) Close() error {
 	prometheus.Unregister(createBoltCollector(s.db))
+
+	// Before DB closes, we should dump the cached state summary objects to DB.
```
Comment (terencechain) [inline beacon-chain/db/kv/kv.go:151]: review

File: beacon-chain/db/kv/state_summary.go [hunk 3]
```diff
@@ -4,6 +4,7 @@ import (
 	"context"
 
 	pb "github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1"
+	"github.com/prysmaticlabs/prysm/shared/bytesutil"
```
Comment (terencechain) [inline beacon-chain/db/kv/state_summary.go:7]: review this file

File: beacon-chain/state/stategen/migrate.go [hunk 4]
```diff
@@ -106,11 +106,6 @@ func (s *State) MigrateToCold(ctx context.Context, fRoot [32]byte) error {
 		}
 	}
 
-	// Migrate all state summary objects from state summary cache to DB.
```
Comment (terencechain) [inline beacon-chain/state/stategen/migrate.go:109]: This functionality is now built within DB

File: beacon-chain/db/kv/kv.go (hunk 2 above)
Comment (rauljordan) [reply beacon-chain/db/kv/kv.go:151]: We should add a context to this. Let's refactor the function signature of NewKVStore

File: beacon-chain/db/kv/state_summary.go [hunk 5]
```diff
@@ -78,3 +89,25 @@ func (s *Store) stateSummaryBytes(ctx context.Context, blockRoot [32]byte) ([]by
...
 	return enc, err
 }
+
+// This saves all cached state summary objects to DB, and clears up the cache.
+func (s *Store) saveCachedStateSummariesDB(ctx context.Context) error {
+	summaries := s.stateSummaryCache.getAll()
+	if err := s.db.Update(func(tx *bolt.Tx) error {
+		bucket := tx.Bucket(stateSummaryBucket)
+		for _, s := range summaries {
+			enc, err := encode(ctx, s)
```
Comment (rauljordan) [inline beacon-chain/db/kv/state_summary.go:99]: Ideally this sort of operation should be done outside of the bolt tx to save on time the db lock is held, but no big deal, just something to keep in mind if this method is slow in the future

Comment (terencechain) [reply beacon-chain/db/kv/state_summary.go:99]: Good point. Moved encoding outside of the tx:
```go
	encs := make([][]byte, len(summaries))
	for i, s := range summaries {
		enc, err := encode(ctx, s)
		encs[i] = enc
	}
	if err := s.db.Update(func(tx *bolt.Tx) error {
		bucket := tx.Bucket(stateSummaryBucket)
		for i, s := range summaries {
			bucket.Put(s.Root, encs[i])
		}
	});
```

**Review by prestonvanloon — CHANGES_REQUESTED** [review]

File: beacon-chain/operations/slashings/service.go [hunk 6]
```diff
@@ -130,7 +130,7 @@ func (p *Pool) InsertAttesterSlashing(
 	slashedVal := sliceutil.IntersectionUint64(slashing.Attestation_1.AttestingIndices, slashing.Attestation_2.AttestingIndices)
 	cantSlash := make([]uint64, 0, len(slashedVal))
 	for _, val := range slashedVal {
-		// Has this validator index been included recently?
+		// has this validator index been included recently?
```
[the remark this answers is not in the document]
Comment (terencechain) [reply beacon-chain/operations/slashings/service.go]: Goland's refactor feature decided to change comment when I un-exported state summary's has method

It's ready again

**Review by prestonvanloon — APPROVED** [review]

**Pull request merged**